accepted/00065-salt-minions-update-cycle.md

```diff
@@ -0,0 +1,127 @@
+- Ensure identical version of Salt Minions with the Salt Master.
+- Independence from client OS's maintenance cycle.
+
+# Detailed design
+[design]: #detailed-design
+
+## Approach
+
+**Independent Maintenance Cycle**
+
+In order to efficiently update Salt Minion on any version of operating system, the update cycle of a Salt Minion needs to be independent from the maintenance update of the very operating system itself.
+
+Rationale: Allow components update as soon as it is needed for the software component, without waiting for the general maintenance update.
+
+**Isolation**
+
+Since Salt Minion is a part of configuration management system, it is also a good to have it "static", in terms of zero dependencies from the operating system it currently runs.
+
+Rationale: Allow running latest modern software on much older environments without any impact to the managed operating system.
+
+**In-place Bootstrapping**
+
+Each Salt Minion is not packaged anymore in OS-specific package manager (RPM, Deb etc), but is bootstrapped via standard Python mechanism (Package Installer for Python, PIP).
+
+Rationale: Track all the patches in only one place. As the code is isolated (see above), avoid unnecessary extra-packaging for isolated version of the Salt Minion.
+
+## Principle
+
+The principle is to run own minimal [PyPi](https://pypi.org/project/pypiserver/) local server on Uyuni cluster. In order to install a Salt Minion, a managed system should use PIP againstlocal PyPi server, as shown on Figure 1 below:
+
+![Layout](images/00065-layout.png)
+
+_Figure 1_
+
+Process update or installation flow as follows:
+1. One RPM general package of Salt Minion **(1)** is provided for all possible platforms and architectures and is installed on PyPi server. However, Salt Minion is not _installed_ per se in the sense as for end-user, but as a PyPi package into the PyPi repository location. This way patches are still tracked at RPM `.spec` level.
+2. General statically compiled Python vanilla environment is provided as an RPM package from Uyuni repository and is allowed to be installed **(2)** on a Managed System, e.g. in `/opt/python3.8` or whatever version is needed. That way it is also allowed to have many Python versions as well as its modifications, if ever needed.
+3. Once virtual environment (not shown on the _Figure 1_) is created for the the separate Python interpreter, PIP is bootstrapping the entire Salt Minion against Uyuni-served PyPi server with all the required Salt Minion dependencies.
+
+This should result into running fully featured Salt Minion of latest version in an isolated virtual environment on an OS-isolated Python interpreter, which is aligned with the current Uyuni Server. Once Uyuni Server is updated to another version, so is the PyPi container image is also upgraded. Thus, Uyuni Server should issue mass-upgrade to all Salt Minions that should PIP-upgrade themselves, bumping up to the latest available version.
+
```

Comment:
How is this different from when a normal SLE product is updated? First update the SMT repositories, zypper up the main node, restart the services in the main node, go one by one (or via an salt / chef) to update the workers.

IIUC you are only replacing zypper up / apt-get update, with something more brittle: `pip install --upgrade <package-list>`. The order of this package list can break your venv (or site-packages)

Reply:
It is different that you don't have to generate another tons of RPM/deb packages that installing things separately in static way. You have to take care only of installation sources that are deployed on PyPi server. But is it brittle or not robust — isn't it just a technical details? As PoC of pip-boostrapped Salt can be found [here](https://github.com/saltstack/salt-bootstrap/blob/develop/bootstrap-salt.sh), although it seems used in production.

Reply:
> It is different that you don't have to generate another tons of RPM/deb packages that installing things separately in static way.

In that case I did misunderstand the proposal. As I did understand, you are proposing to create the package in OBS, and transform it later to something that can be registered in a local PyPi, so no patch get lost.

If this is correct, yes, you indeed need to generate another tons of RPM/deb packages, together with a way of making them installable with PIP.

This PoC is not answering how the live-cycle of those wheels is managed, that was my question.

Later I attached a proposal of creating in OBS wheels that can be installed in PIP. It is interesting, but is using this approach: you first create the RPM / DEB and OBS takes care of generating the PIP package.

But there is no escape of repackaging all for the patches and the support.

So my question remains: as you already have a bunch of packages, and a distribution channel (SMT), how this proposal is different?

Please, unresolve this comment.